### ISSUE TITLE
Early init call

### ISSUE THREAD

**Issue by rastreus** [SelectViewData/elmish-store#1]
First of all, thank you for this implementation and the [explanation article](https://dev.to/lkrzywizna/optimizing-f-and-react-integration-with-elmish-store-a-guide-to-efficient-state-management-316m). I appreciate your work and any additional contributions to the Fable Elmish community.

It makes sense once you think about it, but I ran into it so I thought I would share since you asked:

> I encourage you to try this approach with your Elmish applications and share your results ... I'm eager to hear about your experiences.

Creating the `store` value calls `init` at the time that it is created.

```fsharp
let store : ElmishStore<Model, Msg> =
        // A placeholder function for view is used because it isn't needed
        Program.mkProgram init update (fun _ _ -> Html.none)
        #if !MOCK // If not mocking (msw), then subscribe to the WebSocket
        |> Program.withSubscription WebSocket.subscribe
        #endif
        #if DEBUG
        |> Program.withDebuggerCoders encoder decoder
        #endif
        // https://github.com/elmish/elmish/blob/40d38cc5c44d58c842385ed6d608e8d97c2ef399/src/program.fs#L144
        |> Program.map
            (fun init -> init)
            (fun update -> update)
            (fun _ -> (fun _ _ -> ())) // map `Fable.React.ReactElement` to `unit`
            (fun setState -> setState)
            (fun subscribe -> subscribe)
            (fun termination -> termination)
        |> ElmishStore.createStore "core"
```

Rather than a value, I previously had this as a function (`let mkProgram () : Program<unit, Model, Msg, unit> =`) that was called in the root component with **Feliz.UseElmish**:

```
let model, dispatch =
    React.useElmish mkProgram
```

Normally, my MSW would get deferred and would be available during init (`fetchDataCmd` function), but I needed to set my init to `Cmd.none` and then call my initial fetch with `React.useEffectOnce`:

```fsharp
module Main

// Entry point must be in a separate file
// for Vite Hot Reload to work

#if DEBUG
open Feliz // For React.strictMode
#endif
open Browser
open Fable.React
open Fable.Core
open Fable.Core.JsInterop
open Router
open Util

importSideEffects "./index.css"

#if DEBUG
[<Import("deferRender", "./mswSetup.js")>]
let deferRender : unit -> JS.Promise<unit> = jsNative
#endif

"root"
|> document.getElementById
|> ReactDomClient.createRoot
|> fun root ->
    #if DEBUG
    deferRender()
    |> Promise.map (fun _ ->
            React.strictMode [
                Router()
                |> toReact
            ]
            |> root.render
        )
    |> Promise.start
    #else
    Router()
    |> toReact
    |> root.render
    #endif
```

```fsharp
[<ReactComponent>]
let App () : Fable.React.ReactElement =

    let fetchingData : bool =
        useSelector (_.FetchingData)

    React.useEffectOnce (fun _ ->
        promise {
            if fetchingData then
                do! Promise.sleep 1000
                FetchData |> dispatch
            else
                ()
        }
        |> Promise.start
    )

    if fetchingData then

        Html.div [
            prop.classes [
                "bg-stone-50 w-screen h-screen"
                "flex justify-center items-center"
            ]
            prop.children [
                Html.span [
                    prop.className "loading loading-spinner loading-lg bg-amber-600"
                ]
            ]
        ]
    else
        // Render UI ...
```

Other than that, everything seems to be working well. Thanks again! 😊

**Comment by lukaszkrzywizna** [comment]
Thank you, @rastreus, for the feedback.

If I understand correctly, you're setting up mocked responses in a debug configuration, which must be completed before React and Elmish initialization. The challenge arises because the model initializes "too fast", not waiting for the mocking promise to resolve. Since Elmish.Store essentially serves as a plugin to add a subscription registry for React hooks and runs the Elmish program, this issue might be more inherent to Elmish itself. I see two potential solutions:

### 1. Defer Elmish-Store Creation

The store could be initialized inside a Promise.map to synchronize the operations. However, this approach requires handling the potential uninitialized state of the store in selectors, which should be manageable given that React won't run before initialization. If issues arise, the compiler may not provide assistance.

```fsharp
let mutable store = None

let initializeStore () =
  let main =
      Program.mkProgram init update (fun _ _ -> ())
      |> ElmishStore.createStore "main"
  store <- Some main

[<Hook>]
let useSelector (selector: Model -> 'a) = React.useElmishStore (store.Value, selector)

[<Hook>]
let useSelectorMemoized (memoizedSelector: Model -> 'a) =
  React.useElmishStoreMemoized (store.Value, memoizedSelector)

let dispatch msg = match store with | Some store -> store.Dispatch msg | None -> failwith "Store not initialized"

let myPromise () = Promise.sleep 1000

myPromise ()
|> Promise.map (fun _ ->
    ModelStore.initializeStore()
    ReactDOM
      .createRoot(Browser.Dom.document.getElementById "elmish-app")
      .render (React.strictMode [ View.AppView() ])
)
|> Promise.start
```
This approach, however, doesn't fully support HMR as detailed in [Issue #2](https://github.com/SelectViewData/elmish-store/issues/2), potentially requiring a custom implementation to avoid reinitializing the store on every update of init/update functions.

### 2. Create a Listener for Mock Solution Loading
This can be achieved either through:

- **Elmish subscription:** Likely the safest option. This method involves setting a flag once loading is complete and using an Elmish subscription to trigger actions accordingly.

```fsharp
let mutable isLoaded = ref false

let notifyIfLoaded dispatch =
    let handler =
        JS.setInterval
            (fun () -> if isLoaded.Value then dispatch FetchData)
            1000
    React.createDisposable (fun () -> JS.clearInterval handler)

let subscribe _model =
    if isLoaded.Value then
        []
    else
        [ ["sync"], notifyIfLoaded ]

let store =
  Program.mkProgram init update (fun _ _ -> ())
  |> Program.withSubscription subscribe
  |> ElmishStore.createStore "main"
  
let myPromise () = Promise.sleep 1000

myPromise ()
|> Promise.map (fun _ ->
    ModelStore.isLoaded.Value <- true
    ReactDOM
      .createRoot(Browser.Dom.document.getElementById "elmish-app")
      .render (React.strictMode [ View.AppView() ])
)
|> Promise.start
```

- **React hook**: A simpler method, similar to your current implementation, using React hooks to manage the loading state and trigger actions post-load.

By focusing on these two approaches, we can accommodate the need for asynchronous setup operations without significantly altering the core behavior of `Elmish.Store`.

**Comment by lukaszkrzywizna** [comment]
Since there is no response, I guess that everything is clear. Feel free to reopen if you need to discuss it further 🙂
